commands/key_add.go

```diff
@@ -0,0 +1,89 @@
+package commands
+
+import (
+	"fmt"
+
+	"github.com/MichaelMure/git-bug/cache"
+	"github.com/MichaelMure/git-bug/commands/select"
+	"github.com/MichaelMure/git-bug/identity"
+	"github.com/MichaelMure/git-bug/input"
+	"github.com/MichaelMure/git-bug/util/interrupt"
+	"github.com/spf13/cobra"
+)
+
+var (
+	keyAddArmoredFile    string
+	keyAddArmored        string
+	keyAddExpireUnixTime int64
+)
+
+func runKeyAdd(cmd *cobra.Command, args []string) error {
+	backend, err := cache.NewRepoCache(repo)
+	if err != nil {
+		return err
+	}
+	defer backend.Close()
+	interrupt.RegisterCleaner(backend.Close)
+
+	id, args, err := _select.ResolveUser(backend, args)
```

Comment:
you should only be able to add a key on a user you are allowed to control. There is two scenario where this is the case:
- the identity has no key yet ("unprotected")
- you have the private key of one of the public key present in the last version

This rules should be implemented in the `Validate()` function of Identity so they are checked 1) before a commit and 2) when merging remote data.

Reply:
If the user loses the private key, the identity can't be used anylonger. 

(In our case, we trust everybody to add/remove keys to anybody. The server ("when merging remote data"?) would run a custom check to make sure the keyring is valid and that everything originates from one key.)

Reply:
Is there a notion of "admin identity"?

Reply:
> If the user loses the private key, the identity can't be used any longer.

Yes, that's the kind of hard rules you end up having with distributed applications. Some sort of backup or trust system could be imagined but for now that looks like the good way to go to me.

> The server ("when merging remote data"?) would run a custom check to make sure the keyring is valid and that everything originates from one key.)

Why do you need code server side ? I understand you have different requirements, but is there a way for you to implement those extra requirements while staying close to git-bug's design as a distributed app ? It would help both of us if we can manage that.

> Is there a notion of "admin identity"?

Not yet but that would be nice. To reach that point the plan is:
- have a shared project config data structure (valid labels , .... )
- store in that new structure who is admin, with the mechanism to update it
- implement this admin role as rules to check when accepting new data (bugs + project config at least)

Reply:
> > If the user loses the private key, the identity can't be used any longer.
> 
> Yes, that's the kind of hard rules you end up having with distributed applications. Some sort of backup or trust system could be imagined but for now that looks like the good way to go to me.

What would be the problem with trusting everybody to make changes to identities?

> > The server ("when merging remote data"?) would run a custom check to make sure the keyring is valid and that everything originates from one key.)
> 
> Why do you need code server side ? I understand you have different requirements, but is there a way for you to implement those extra requirements while staying close to git-bug's design as a distributed app ? It would help both of us if we can manage that.

The problem is the users can push anything working around the local validation done when they use git-ticket. The pre-receive server-side git hook prevents merging bad changes.

Reply:
> What would be the problem with trusting everybody to make changes to identities?

In the general case people that have a write access to a git repo are trustworthy but that's not always the case. Git support signature to be able to prove that someone made something and improve the trust you can have in the system, but if someone is able to change the valid keys of anyone else, this signature system is worth nothing.

In a similar fashion, this identity system will be the base of the role management. Some users will be admin of a project (or more complex roles) and some not. If you can tamper the keys of anyone, this system is equally useless.

If everyone is trustworthy it's all good and you don't even have to bother with keys and you can leave your identity unprotected. When it's not the case, you should be protected from abuse.

> The problem is the users can push anything working around the local validation done when they use git-ticket. The pre-receive server-side git hook prevents merging bad changes.

Why is that the case ? If your user use `git ticket pull` (which is much simpler than manually merging remote hidden branches), data will be checked. Everything that doesn't follow the rules will be discarded.

My point is, even if a **malicious** user manage to push bad changes, those changes will not propagate. A non-malicious user should not be able to create bad data as it is checked before writing to git.

Now I agree that its not perfect and a server side hook to control the data is useful. Would you share the one you have ?

Reply:
Huhu sorry, reading myself again, I realize I rewrote what you were saying.

Reply:
The signatures are useful to prove the changes have not been made by an authorized user pretending to be some other authorized user.

The valid keys can be changed by any authorized user, but it will be obvious who changed them, since the server shall not accept forced pushes. 

The malicious authorized users can work around the local validation, but they shall not be allowed to push, because we can. :) Yes, I'll share the validator!

Reply:
> The valid keys can be changed by any authorized user, but it will be obvious who changed them, since the server shall not accept forced pushes.

Yes, you can do forensics to find out who dropped the ball, but it would be better to actually prevent it, no ?

Reply:
Thinking out loud, there should probably be a `Keystore` interface somewhere with multiple implementations:
- git config
- gpg
- some centralized system that allow recovery for "corporate" environment

Maybe even an implementation that allow to have multiple sub-keystore for example git config + gpg.